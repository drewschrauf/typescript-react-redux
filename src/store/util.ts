import * as _ from 'lodash';

export interface IAction<T> {
  readonly type: string;
  readonly payload: T;
}

interface IActionCreator<T> {
  readonly type: string;
  (payload: T): IAction<T>;
}

export const createAction = <T extends {}>(type: string): IActionCreator<T> =>
  _.assign((payload: T): any => ({ type, payload }), {
    type,
  });

export const isType = <T extends {}>(
  action: IAction<any>,
  actionCreator: IActionCreator<T>,
): action is IAction<T> => action.type === actionCreator.type;
