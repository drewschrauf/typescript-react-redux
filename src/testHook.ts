const testHook = (name: string) =>
  process.env.NODE_ENV !== 'production' ? { 'data-test': name } : {};
export default testHook;

export const selectorForTestHook = (name: string, element: string = '') =>
  `${element}[data-test="${name}"]`;
