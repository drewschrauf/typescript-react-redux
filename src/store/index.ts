import { applyMiddleware, combineReducers, compose, createStore } from 'redux';
import thunk, { ThunkDispatch } from 'redux-thunk';
import counter, { ICounterState } from './reducers/counter';
import { IAction } from './util';

const reducer = combineReducers({
  counter,
});

export interface IState {
  counter: ICounterState;
}

export type Dispatch = ThunkDispatch<IState, any, IAction<any>>;

const store = createStore(
  reducer,
  compose(
    applyMiddleware(thunk),
    (window as any).__REDUX_DEVTOOLS_EXTENSION__ && (window as any).__REDUX_DEVTOOLS_EXTENSION__(),
  ),
);

export default store;
