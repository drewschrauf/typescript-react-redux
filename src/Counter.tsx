import * as React from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";

import { IState } from "./store";
import { decrementBy, delayedIncrement, incrementBy } from "./store/actions/counter";

interface IAppProps { } // tslint:disable-line:no-empty-interface

interface IConnectedState {
  count: number;
  pending: boolean;
}

interface IConnectedDispatch {
  increment: () => void;
  decrement: () => void;
  delayedIncrement: () => void;
}

const AppComponent = (props: IAppProps & IConnectedState & IConnectedDispatch) => (
  <div>
    <h1>Count {props.count}</h1>
    <button onClick={props.increment}>Increment</button>
    <button onClick={props.decrement}>Decrement</button>
    <button disabled={props.pending} onClick={props.delayedIncrement}>Delayed increment</button>
  </div>
);

const App: React.ComponentClass<IAppProps> = connect(
  (state: IState, ownProps: IAppProps): IConnectedState => ({
    count: state.counter.count,
    pending: state.counter.pending,
  }),
  (dispatch: Dispatch<IState>): IConnectedDispatch => ({
    decrement: () => dispatch(decrementBy({ amount: 1 })),
    delayedIncrement: () => dispatch(delayedIncrement({ amount: 1 })),
    increment: () => dispatch(incrementBy({ amount: 1 })),
  }),
)(AppComponent);
export default App;
