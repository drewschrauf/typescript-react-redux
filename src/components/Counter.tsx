import React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { State, Dispatch } from '../store';
import { decrementBy, delayedIncrement, incrementBy } from '../store/actions/counter';
import testHook from '../testHook';

interface CounterProps {
  /** The amount to increment the counter by on each click */
  readonly incrementAmount: number;
}

interface ConnectedState {
  /** The current count */
  readonly count: number;
  /** Is a delayed increment currently being performed? */
  readonly pending: boolean;
}

interface ConnectedDispatch {
  /** Action to perform when clicking increment button */
  readonly increment: () => void;
  /** Action to perform when clicking decrement button */
  readonly decrement: () => void;
  /** Action to perform when clicking delayed increment button */
  readonly delayedIncrement: () => void;
}

const Wrapper = styled.div`
  padding: 10px;
  border: 1px dashed black;
`;

const Button = styled.button`
  margin-right: 10px;
  font-family: inherit;

  width: 100%;
  margin-bottom: 5px;
  @media screen and (min-width: 450px) {
    width: initial;
    margin-bottom: 0;
  }
`;

const CounterComponent = ({
  incrementAmount,
  count,
  increment,
  decrement,
  pending,
  delayedIncrement: incrementWithDelay,
}: CounterProps & ConnectedState & ConnectedDispatch) => (
  <Wrapper>
    <h1 {...testHook('count')}>Count {count}</h1>
    <Button {...testHook('increment')} onClick={increment}>
      Increment by {incrementAmount}
    </Button>
    <Button {...testHook('decrement')} onClick={decrement}>
      Decrement by {incrementAmount}
    </Button>
    <Button {...testHook('delayed-increment')} disabled={pending} onClick={incrementWithDelay}>
      Delayed increment by {incrementAmount}
    </Button>
  </Wrapper>
);

/**
 * React component that renders the main UI. It displays the current count and provides buttons
 * for modifying it.
 */
const Counter = connect(
  (state: State, _ownProps: CounterProps) => ({
    count: state.counter.count,
    pending: state.counter.pending,
  }),
  (dispatch: Dispatch, ownProps: CounterProps) => ({
    decrement: () => dispatch(decrementBy({ amount: ownProps.incrementAmount })),
    delayedIncrement: () => dispatch(delayedIncrement({ amount: ownProps.incrementAmount })),
    increment: () => dispatch(incrementBy({ amount: ownProps.incrementAmount })),
  }),
)(CounterComponent) as React.ComponentClass<CounterProps>;
export default Counter;
