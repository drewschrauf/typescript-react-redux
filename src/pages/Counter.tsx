import React from 'react';
import { RouteComponentProps } from 'react-router-dom';
import { isNaN } from 'lodash';

import Counter from '../components/Counter';

const getIncrementAmount = (by: any) => {
  const value = parseInt(by, 10);
  return !isNaN(value) ? value : 1;
};

const CounterPage = ({ match }: RouteComponentProps) => (
  <Counter incrementAmount={getIncrementAmount((match.params as any).by)} />
);
export default CounterPage;
