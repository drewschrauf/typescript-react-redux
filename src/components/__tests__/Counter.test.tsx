import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore, { MockStoreCreator } from 'redux-mock-store';
import thunk from 'redux-thunk';
import set from 'lodash/fp/set';
import { selectorForTestHook } from '../../testHook';

import Counter from '../Counter';
import { State } from '../../store';

const COUNT_SELECTOR = selectorForTestHook('count');
const INCREMENT_BUTTON_SELECTOR = selectorForTestHook('increment', 'button');
const DECREMENT_BUTTON_SELECTOR = selectorForTestHook('decrement', 'button');
const DELAYED_INCREMENT_BUTTON_SELECTOR = selectorForTestHook('delayed-increment', 'button');

enum TestActionTypes {
  incrementBy,
  decrementBy,
  delayedIncrement,
}

jest.mock('../../store/actions/counter', () => ({
  incrementBy: (...args: any[]) => ({ type: TestActionTypes.incrementBy, args }),
  decrementBy: (...args: any[]) => ({ type: TestActionTypes.decrementBy, args }),
  delayedIncrement: (...args: any[]) => ({ type: TestActionTypes.delayedIncrement, args }),
}));

const defaultState: State = { counter: { count: 1, pending: false } };
const defaultProps = { incrementAmount: 1 };

describe('Counter', () => {
  let mockStore: MockStoreCreator<State>;
  beforeEach(() => {
    mockStore = configureStore<State>([thunk]);
  });

  it('should match the snapshot', () => {
    const store = mockStore(defaultState);
    const root = mount(
      <Provider store={store}>
        <Counter {...defaultProps} />
      </Provider>,
    );
    expect(root.html()).toMatchSnapshot('default render');
  });

  it('should render the current count', () => {
    const store = mockStore(defaultState);
    const root = mount(
      <Provider store={store}>
        <Counter {...defaultProps} />
      </Provider>,
    );
    expect(root.find(COUNT_SELECTOR).text()).toBe('Count 1');
  });

  it('should increment when increment button clicked', () => {
    const store = mockStore(defaultState);
    const root = mount(
      <Provider store={store}>
        <Counter {...defaultProps} />
      </Provider>,
    );
    const handler = root.find(INCREMENT_BUTTON_SELECTOR).prop('onClick');

    (handler as any)();
    expect(store.getActions()).toEqual([
      { type: TestActionTypes.incrementBy, args: [{ amount: 1 }] },
    ]);
  });

  it('should decrement when decrement button clicked', () => {
    const store = mockStore(defaultState);
    const root = mount(
      <Provider store={store}>
        <Counter {...defaultProps} />
      </Provider>,
    );
    const handler = root.find(DECREMENT_BUTTON_SELECTOR).prop('onClick');

    (handler as any)();
    expect(store.getActions()).toEqual([
      { type: TestActionTypes.decrementBy, args: [{ amount: 1 }] },
    ]);
  });

  it('should delay increment when delay increment button clicked', () => {
    const store = mockStore(defaultState);
    const root = mount(
      <Provider store={store}>
        <Counter {...defaultProps} />
      </Provider>,
    );
    const handler = root.find(DELAYED_INCREMENT_BUTTON_SELECTOR).prop('onClick');

    (handler as any)();
    expect(store.getActions()).toEqual([
      { type: TestActionTypes.delayedIncrement, args: [{ amount: 1 }] },
    ]);
  });

  it('should enable delayed increment by default', () => {
    const store = mockStore(defaultState);
    const root = mount(
      <Provider store={store}>
        <Counter {...defaultProps} />
      </Provider>,
    );
    expect(root.find(DELAYED_INCREMENT_BUTTON_SELECTOR).prop('disabled')).toBe(false);
  });

  it('should disable delayed increment if pending', () => {
    const store = mockStore(set('counter.pending', true, defaultState));
    const root = mount(
      <Provider store={store}>
        <Counter {...defaultProps} />
      </Provider>,
    );
    expect(root.find(DELAYED_INCREMENT_BUTTON_SELECTOR).prop('disabled')).toBe(true);
  });
});
