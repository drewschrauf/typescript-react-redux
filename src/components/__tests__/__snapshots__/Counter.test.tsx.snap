// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter should match the snapshot: default render 1`] = `"<div class=\\"sc-bdVaJa cJwtzi\\"><h1 data-test=\\"count\\">Count 1</h1><button data-test=\\"increment\\" class=\\"sc-bwzfXH QqPdO\\">Increment by 1</button><button data-test=\\"decrement\\" class=\\"sc-bwzfXH QqPdO\\">Decrement by 1</button><button data-test=\\"delayed-increment\\" class=\\"sc-bwzfXH QqPdO\\">Delayed increment by 1</button></div>"`;
